nasa/ow_autonomy
# Support for unified pan/tilt telemetry

### Summary of Changes

- Implemented PLEXIL support for the topic `/owl_msgs/pan_tilt_position` as specified in Google doc.
   - Note that this topic is only used for OWLAT Sim.  OceanWATERS still reads pan and tilt from the `/joint_states` topic, because this is more efficient.  (The unified topic republishes this information and adds some lag).
- PLEXIL now provides both radian and degree versions of pan/tilt lookups.
    - Note that the PLEXIL interface is independent of the ROS interface and still reads pan and tilt individually, which is most useful.

### Unrelated Changes

- Added `plans/unified` as a home for PLEXIL plans that can run on either testbed.
- Discovered and worked around a pre-existing bug in Plexil plan selection: a plan that finishes very quickly will break the interface because of synchronization issues.  More info in the code comments.
- Replaced the individual subscriber members with a container.
 - Factored shared members in OwInterface and OwlatInterface into PlexilInterface.
 - Added `const` to many member functions that are.
 - Formatting: removed trailing whitespace in some files (done automatically by Emacs), shortened some lines.

### Test

Two tests are needed, one using OceanWATERS, and one using OWLAT Sim.  We assume you have OWLAT Sim installed.  First, a clean build that supports OWLAT is needed:

```
$ catkin clean
$ catkin build -DOWLAT=ON
```

Test 1: OceanWATERS
1. Start any simulation world.
2. Launch `ow_exec.launch` and run the plan `TestTelemetry`.  You should see something close to the printout below.  Note that we're reading the pan/tilt positions at startup.  Feel free to move the antenna and run the plan again.
3. To test the changes to the plan selection interface, run any other plan, and then run `TestTelemetry` again.
```
[ INFO]: PLEXIL: Pan degees: -0.000373310720624672
[ INFO]: PLEXIL: Pan radians: -6.51550120789324e-06
[ INFO]: PLEXIL: Tilt degees: -0.665347767210253
[ INFO]: PLEXIL: Tilt radians: -0.0116125092086978
```

Test 2: OWLAT Sim
1. Start `roscore`.
2. Start OWLAT Sim: `launch_owlat Scoop`
3. Launch `owlat_exec.launch`
4. Run the same plan, `TestTelemetry`.  Since nothing is publishing to the topic, you should see the Plexil adapter default values:

```
[ INFO]: PLEXIL: Pan degees: 0
[ INFO]: PLEXIL: Pan radians: 0
[ INFO]: PLEXIL: Tilt degees: 0
[ INFO]: PLEXIL: Tilt radians: 0
```
5. Now, start publishing to the topic manually:
```
$ rostopic pub -r 10 /owl_msgs/pan_tilt_position owl_msgs/PanTiltPosition '{ header: {seq: 10, stamp: {secs: 1, nsecs: 1}, frame_id: "3"}, value: [1.0, 2.0] }'
```
6. Run the `TestTelemetry` plan again, and you should see this:
```
[ INFO]: PLEXIL: Pan degees: 57.2957795130823
[ INFO]: PLEXIL: Pan radians: 1
[ INFO]: PLEXIL: Tilt degees: 114.591559026165
[ INFO]: PLEXIL: Tilt radians: 2
```
